doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1588,39 +1588,70 @@ Controlling the proof flow
-   concrete terms from :n:`@one_term_with_bindings`.
-
-   .. todo PR:
-      1) what is "non-dependent implication" above?
-      2) In this little example, what exactly is (H x)? It looks almost like
-         a function but H doesn't look like a function.  Is this just something
-         wierd about specialize or have I missed some basic point?
-
-      Goal forall x: nat, (forall n, n >= 0) -> True.
-      intros.  (* H is forall n : nat, n >= 0 *)
-      specialize (H x). (* H is x >= 0 *)
-
-   The application to :n:`@one_term` can be partial. The
-   uninstantiated arguments are inferred by unification if possible or left
-   quantified in the hypothesis otherwise. With the :n:`as` clause, the local
-   hypothesis :n:`@ident` is left unchanged and instead, the modified hypothesis
-   is introduced as specified by the :token:`simple_intropattern`. The name :n:`@ident`
-   can also refer to a global lemma or hypothesis. In this case, for
-   compatibility reasons, the behavior of :tacn:`specialize` is close to that of
-   :tacn:`generalize`: the instantiated statement becomes an additional premise of
-   the goal. The ``as`` clause is especially useful in this case to immediately
-   introduce the instantiated statement as a local hypothesis.
+   If the :term:`head constant` of :n:`@one_term` has the type
+   `forall ...`, the tactic replaces one or more of the
+   quantified variables in the type with terms provided by
+   :n:`@one_term_with_bindings`, either in the form of a
+   :ref:`function application <function_application>` (which may be partial),
+   such as `(H 1)`, or with named binders, such as `H with (n:=1)`.
+
+   If the :term:`head constant` has a non-dependent product type such as
+   `A -> B -> C`, the tactic does :term:`forward reasoning`, eliminating one or
+   more of the premises.
+
+   Uninstantiated arguments are inferred by unification, if possible, or otherwise
+   left quantified in the resulting term.
+
+   If the :term:`head constant` is a hypothesis, the resulting term replaces that
+   hypothesis.  Specifying :n:`@as_ipat` will leave the original hypothesis
+   unchanged and will introduce new hypotheses as specified by the
+   :token:`simple_intropattern`.
+
+   If the head constant is a lemma or theorem, the resulting term
+   is added as a new premise of the goal so that the behavior is similar
+   to that of :tacn:`generalize`.  In this case, you can use :n:`@as_ipat` to
+   immediately introduce the modified term as one or more hypotheses.
 
    .. exn:: @ident is used in hypothesis @ident.
       :undocumented:
 
    .. exn:: @ident is used in conclusion.
       :undocumented:
 
+   .. example:: partial application in :tacn:`specialize`
+
+      .. coqtop:: reset none
+
+         Goal (forall n m: nat, n + m = m + n) -> True.
+
+      .. coqtop:: out
+
+         intros.
+
+      .. coqtop:: all
+
+         specialize (H 1). (* equivalent to: specialize H with (n := 1) *)
```

Comment:
Maybe add an example illustrating my remark above (picking which parameters to instantiate)?